O3-1409: Support for users changing passwords

jnsereko opened the pull request:
# Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. Ensure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing#contributing-guidelines) label (such as `feat`, `fix`, or `chore`, among others). See existing PR titles for inspiration.

## For changes to apps

- [x] My work conforms to the [**O3 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://om.rs/o3ui).

## If applicable

- [x] My work includes tests or is validated by existing tests.
- [x] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary
<!-- Please describe what problems your PR addresses. -->
 O3 screen to allow users to change their password.
cc @michaelbontyes @vasharma05 @denniskigen @hadijahkyampeire 

## Screenshots
<!-- Required if you are making UI changes. -->

https://github.com/openmrs/openmrs-esm-core/assets/58003327/1c41012c-97a6-4d5e-b0d7-d31a8cbb00af


## Related Issue
<!-- Paste the link to the Jira ticket here if one exists. -->
<!-- https://issues.openmrs.org/browse/O3- -->
https://openmrs.atlassian.net/browse/O3-1409

## Other
<!-- Anything not covered above -->

ibacher commented on packages/apps/esm-login-app/src/change-password-button/change-password-button.component.tsx:
```diff
@@ -0,0 +1,32 @@
+
+  return (
+    <>
+      <SwitcherDivider className={styles.divider} />
```
I don't think that the `SwitcherDivider` should be part of this class.

ibacher commented on packages/apps/esm-login-app/src/change-password-button/change-password-button.component.tsx:
```diff
@@ -0,0 +1,32 @@
+  );
+};
+
+export default ChangePasswordLink;
```
```suggestion
export default ChangePasswordLink;

```

ibacher commented on packages/apps/esm-login-app/src/change-password-button/change-password-button.scss:
```diff
@@ -0,0 +1,8 @@
+  padding-right: 0rem;
+  @include brand-02(background-color);
+  @extend .productiveHeading01;
+  width: 16rem;
```
Using the Carbon `$spacing` utils. I'm also not sure why you're importing `../root.scss` here?

ibacher commented on packages/apps/esm-login-app/src/change-password/change-password.component.tsx:
```diff
@@ -0,0 +1,232 @@
+      const minLengthPassword = minLengthRegExp.test(passwordInputValue);
+      let errMsg = '';
+      if (passwordLength === 0) {
+        errMsg = 'Password is empty';
```
Any display string always needs to be translatable.

ibacher commented on packages/apps/esm-login-app/src/change-password/change-password.component.tsx:
```diff
@@ -0,0 +1,232 @@
+                    {t('discard', 'Discard')}
+                  </Button>
+                  <Button
+                    style={{ maxWidth: '50%' }}
```
I'd use SCSS rather than inline styles for this.

ibacher commented on packages/apps/esm-login-app/src/change-password/change-password.resource.ts:
```diff
@@ -0,0 +1,19 @@
+import { openmrsFetch } from '@openmrs/esm-framework';
+
+export async function performPasswordChange(oldPassword: string, newPassword: string) {
+  const abortController = new AbortController();
```
Not much use for an AbortController inside this method. This is what `useAbortController()` is for (automatically aborts requests when the UI unmounts a component).

ibacher commented on packages/apps/esm-login-app/src/change-password/change-password.component.tsx:
```diff
@@ -0,0 +1,232 @@
+      try {
+        setIsSavingPassword(true);
+        const response = await performPasswordChange(passwordInput.oldPassword, passwordInput.confirmPassword);
+        if (response.ok) {
```
`if (repsonse.ok)` is redundant. `openmrsFetch` throws an exception (so the Promise will throw an error) if the result isn't ok.

ibacher commented on packages/apps/esm-login-app/src/change-password/change-password.component.tsx:
```diff
@@ -0,0 +1,232 @@
+        const response = await performPasswordChange(passwordInput.oldPassword, passwordInput.confirmPassword);
+        if (response.ok) {
+          performLogout().then(() => {
+            const defaultLang = document.documentElement.getAttribute('data-default-lang');
```
Uh... just redirect to the logout handler I guess... We've never forced people to re-login after changing their password though, so I'm not a huge fan of this.

ibacher reviewed (changes requested):
@jnsereko This is great work in many ways! A few things:

1. I think it would be preferable to have this as part of the (much neglected) openmrs-esm-admin-tools repo.
2. Instead of a full page, I think the password change could be handled as a modal.
3. As a modal, I think the width could be much smaller than it presently is.

jnsereko commented:
thank you @ibacher for taking a look at this.  
i have moved this logic to esm-admin-tools repo and added it to the `@openmrs/esm-system-admin-app'  hence i am closing this PR 
 follow up PR -> https://github.com/openmrs/openmrs-esm-admin-tools/pull/43
